File: core/src/org/joltsphere/mechanics/MountainClimbingPlayer.java
package org.joltsphere.mechanics;

import org.joltsphere.main.JoltSphereMain;
import org.joltsphere.misc.ObjectData;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.BodyDef;
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType;
import com.badlogic.gdx.physics.box2d.CircleShape;
import com.badlogic.gdx.physics.box2d.FixtureDef;
import com.badlogic.gdx.physics.box2d.World;

public class MountainClimbingPlayer {
	
	@SuppressWarnings("unused")
	private World world;
	public Body body;
	private BodyDef bdef;
	private FixtureDef fdef;
	private CircleShape circle;
	
	private Color color;
	private boolean isGrounded = false;
	private boolean canDouble = false;
	
	private float ppm = JoltSphereMain.ppm;
	@SuppressWarnings("unused")
	private float dt = 0.01666f;
	
	public MountainClimbingPlayer(World world, float x, float y, Color color) {
		this.world = world;
		this.color = color;
		
		bdef = new BodyDef();
		bdef.type = BodyType.DynamicBody;
		bdef.position.set(x / ppm, y / ppm);
		bdef.fixedRotation = false;
		bdef.bullet = true;
		bdef.linearDamping = 0.2f;
		bdef.angularDamping = 0.5f;
		body = world.createBody(bdef);
		body.setUserData(new ObjectData("mountainClimber"));
		fdef = new FixtureDef();
		circle = new CircleShape();
		circle.setRadius(50 / ppm);
		fdef.shape = circle;
		fdef.friction = 0.1f;
		fdef.density = 10;
		fdef.restitution = 0f;
		body.createFixture(fdef);
		
	}
	
	public void shapeRender(ShapeRenderer shapeRender) {
		shapeRender.setColor(color);
		shapeRender.circle(body.getPosition().x*ppm,body.getPosition().y*ppm, 50);
		shapeRender.setColor(Color.GOLD);
	}
	
	public void update(float dt, int groundContacts) {
		this.dt = dt;
		
		if (groundContacts > 0) isGrounded = true; 
		else isGrounded = false;
		
		if (isGrounded) {
			canDouble = true;
		}
	}

	public void moveLeft() {
		moveHorizontal(-1);
	}
	public void moveRight() {
		moveHorizontal(1);
	}
	private void moveHorizontal(int dir) {
		body.applyForceToCenter(30*dir, 0, true);
		body.applyTorque(-5 * dir, true);
	}
	public void jump() {
		if (isGrounded) {
			body.setLinearVelocity(body.getLinearVelocity().x * 0.3f, body.getLinearVelocity().y * 0.3f);
			body.applyLinearImpulse(new Vector2(0, 16), body.getPosition(), true);
		}
		if (canDouble) {
			canDouble = false;
			body.setLinearVelocity(0, 0);
			body.applyLinearImpulse(new Vector2(0, 12), body.getPosition(), true);
		}
		//body.applyForceToCenter(0, 40, true);
	}
	public void fly() {
		body.applyForceToCenter(new Vector2(0, 60), true);
	}

	public void input(int up, int down, int left, int right, int modifier, boolean fly) {	
		if (Gdx.input.isKeyPressed(left)) moveLeft();
		if (Gdx.input.isKeyPressed(right)) moveRight();
		if (Gdx.input.isKeyPressed(up) && fly) fly();
		else if (Gdx.input.isKeyJustPressed(up)) jump();
	}
	
}

File: core/src/org/joltsphere/mechanics/MountainClimbingPlayer.kt
package org.joltsphere.mechanics

import org.joltsphere.main.JoltSphereMain
import org.joltsphere.misc.ObjectData

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Body
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.BodyDef.BodyType
import com.badlogic.gdx.physics.box2d.CircleShape
import com.badlogic.gdx.physics.box2d.FixtureDef
import com.badlogic.gdx.physics.box2d.World

class MountainClimbingPlayer(private val world: World, x: Float, y: Float, private val color: Color) {
    var body: Body
    private val bdef: BodyDef
    private val fdef: FixtureDef
    private val circle: CircleShape
    private var isGrounded = false
    private var canDouble = false

    private val ppm = JoltSphereMain.ppm
    private var dt = 0.01666f

    init {

        bdef = BodyDef()
        bdef.type = BodyType.DynamicBody
        bdef.position.set(x / ppm, y / ppm)
        bdef.fixedRotation = false
        bdef.bullet = true
        bdef.linearDamping = 0.2f
        bdef.angularDamping = 0.5f
        body = world.createBody(bdef)
        body.userData = ObjectData("mountainClimber")
        fdef = FixtureDef()
        circle = CircleShape()
        circle.radius = 50 / ppm
        fdef.shape = circle
        fdef.friction = 0.1f
        fdef.density = 10f
        fdef.restitution = 0f
        body.createFixture(fdef)

    }

    fun shapeRender(shapeRender: ShapeRenderer) {
        shapeRender.color = color
        shapeRender.circle(body.position.x * ppm, body.position.y * ppm, 50f)
        shapeRender.color = Color.GOLD
    }

    fun update(dt: Float, groundContacts: Int) {
        this.dt = dt

        if (groundContacts > 0)
            isGrounded = true
        else
            isGrounded = false

        if (isGrounded) {
            canDouble = true
        }
    }

    fun moveLeft() {
        moveHorizontal(-1)
    }

    fun moveRight() {
        moveHorizontal(1)
    }

    private fun moveHorizontal(dir: Int) {
        body.applyForceToCenter((30 * dir).toFloat(), 0f, true)
        body.applyTorque((-5 * dir).toFloat(), true)
    }

    fun jump() {
        if (isGrounded) {
            body.setLinearVelocity(body.linearVelocity.x * 0.3f, body.linearVelocity.y * 0.3f)
            body.applyLinearImpulse(Vector2(0f, 16f), body.position, true)
        }
        if (canDouble) {
            canDouble = false
            body.setLinearVelocity(0f, 0f)
            body.applyLinearImpulse(Vector2(0f, 12f), body.position, true)
        }
        //body.applyForceToCenter(0, 40, true);
    }

    fun fly() {
        body.applyForceToCenter(Vector2(0f, 60f), true)
    }

    fun input(up: Int, down: Int, left: Int, right: Int, modifier: Int, fly: Boolean) {
        if (Gdx.input.isKeyPressed(left)) moveLeft()
        if (Gdx.input.isKeyPressed(right)) moveRight()
        if (Gdx.input.isKeyPressed(up) && fly)
            fly()
        else if (Gdx.input.isKeyJustPressed(up)) jump()
    }

}
